sync/category.go

```diff
@@ -0,0 +1,93 @@
+package sync
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/appuio/appuio-cloud-reporting/pkg/db"
+	"github.com/vshn/appuio-odoo-adapter/odoo/model"
+)
+
+// Reconcile synchronizes model.InvoiceCategory in Odoo based on the given db.Category according to the following rules:
+//  * If db.Category.Target is NULL then it will search for an existing category that matches the name:
+//    * If not found, it will create a new model.InvoiceCategory and set db.Category.Target to the ID returned by Odoo.
+//    * If found a match, it will use the model.InvoiceCategory and set db.Category.Target to the ID returned by Odoo.
+//  * If db.Category.Target has a value then it will search for a matching model.InvoiceCategory:
+//    * If not found, it will recreate the model.InvoiceCategory.
+//    * If found and model.InvoiceCategory.Name is up-to-date, it will return without error
+//    * If found and model.InvoiceCategory.Name differs from db.Category.Source, the model.InvoiceCategory is updated/reset.
+func (s *InvoiceCategoryReconciler) Reconcile(ctx context.Context, category *db.Category) error {
+	converted := model.InvoiceCategory{}
+	err := CategoryConverter{}.ToInvoiceCategory(category, &converted)
+	if err != nil {
+		return err
+	}
+	if !category.Target.Valid {
+		return s.findOrCreateCategory(ctx, category, converted)
+	}
+	return s.updateCategoryIfNeeded(ctx, category, converted)
+}
+
+func (s *InvoiceCategoryReconciler) findOrCreateCategory(ctx context.Context, category *db.Category, converted model.InvoiceCategory) error {
+	found, err := s.odoo.SearchInvoiceCategoriesByName(ctx, converted.Name)
+	if err != nil {
+		return err
+	}
+	if len(found) == 0 {
```

Comment:
Do we really need the adoption feature?  Is the name unique in Odoo?

Reply:
The name is not unique in Odoo. There could be multiple entries with different IDs and different settings (pagebreak etc) and all sharing the same name.

Reply:
I'd vote for no adoption too. A category does not have that many options so a tiny mess is worth it for simpler code. But lookup by id for sure.